### PR TITLE
NF-560 - ALF integration

### PR DESCRIPTION
**Pull request by SteveSugden**

### CODE REVIEW

File: test/controllers/AgentImporterAddressFrontendControllerSpec.scala [hunk 1]
```diff
@@ -17,57 +17,66 @@
...
 
       application.stop()
     }
 
     "populate the view correctly on a GET when the question has previously been answered" in {
 
       val userAnswers = UserAnswers(userAnswersId).set(
-        AgentImporterManualAddressPage,
-        Address("address line 1", Some("address line 2"), "city", Some("Region"), "GB", "AA211AA")
+        AgentImporterAddressPage,
```
Comment (njmoorewitchcat) [inline test/controllers/AgentImporterAddressFrontendControllerSpec.scala:66]: AMBER: We should also add and test an international address (without a postcode, as above comments).  I'm not expecting the controller to return validation failures for them.

File: test/controllers/ImporterAddressFrontendControllerSpec.scala [hunk 2]
```diff
@@ -17,57 +17,66 @@
...
 
       application.stop()
     }
 
     "populate the view correctly on a GET when the question has previously been answered" in {
 
       val userAnswers = UserAnswers(userAnswersId).set(
-        ImporterManualAddressPage,
-        Address("address line 1", Some("address line 2"), "city", Some("Region"), "GB", "AA211AA")
+        ImporterAddressPage,
```
Comment (njmoorewitchcat) [inline test/controllers/ImporterAddressFrontendControllerSpec.scala:66]: AMBER: Ditto from above, but for the importer controller. "We should also add and test an international address (without a postcode, as above comments). I'm not expecting the controller to return validation failures for them."

File: test/data/TestData.scala [hunk 3]
```diff
@@ -43,10 +51,12 @@ object TestData {
...
   val testRepresentativeAgentName: String            = "Agent name"
   val testRepresentativeImporterName: UserName       = UserName("ImporterName")
-  val testImporterManualAddress: Address             = Address("line 1", Some("line 2"), "City", Some("Region"), "GB", "AA11AA")
+
+  val testImporterManualAddress: Address =
+    Address("line 1", Some("line 2"), "City", Some("Region"), ukCountry, "AA11AA")
 
   val testAgentManualAddress: Address =
-    Address("line 1 agent", Some("line 2 agent"), "City agent", Some("Region agent"), "IT", "AA11AA")
+    Address("line 1 agent", Some("line 2 agent"), "City agent", Some("Region agent"), Country("IT", "Italy"), "AA11AA")
```
Comment (njmoorewitchcat) [inline test/data/TestData.scala:59]: CLARIFICATION: Postcode optional for non GB countries?  Does this effect this line, do we need an additional version or am I barking the wrong tree.

**Review by njmoorewitchcat — CHANGES_REQUESTED** [review]

File: app/controllers/ImporterAddressFrontendController.scala [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  def onChange(): Action[AnyContent] = (identify andThen getData andThen requireData).async {
+    implicit request =>
+      addressLookupService.initialiseJourney(
+        controllers.routes.ImporterAddressFrontendController.onUpdate("").url,
+        controllers.routes.IndexController.onPageLoad().url,
+        controllers.routes.SignOutController.signOut().url,
+        controllers.routes.KeepAliveController.keepAlive().url,
+        if (request.userAnswers.isImporterJourney) "importerYourAddress.title" else "importerManualAddress.title",
+        if (request.userAnswers.isImporterJourney) "importerManualAddress.hint" else ""
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/controllers/ImporterAddressFrontendController.scala]: Checked with design and will add same agent hint text as NIDAC

File: app/forms/AgentImporterManualAddressFormProvider.scala [hunk 5]
```diff
@@ -31,72 +32,84 @@ class AgentImporterManualAddressFormProvider @Inject() extends Mappings {
...
-            )
+  def apply(): Form[Address] = {
+
+    val formToModel = (
+      addressLine1: String,
+      addressLine2: Option[String],
+      city: String,
+      region: Option[String],
+      countryCode: String,
+      postCode: String
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/forms/AgentImporterManualAddressFormProvider.scala:43]: Thank you.  Optional postcode for non-GB totally missed.   Am re-working,

File: app/forms/ImporterManualAddressFormProvider.scala [hunk 6]
```diff
@@ -16,85 +16,96 @@
...
-              regexp(Validation.safeInputPattern, "importerAddress.line1.error.invalid")
-            )
+  def apply(): Form[Address] = {
+    val formToModel = (
+      addressLine1: String,
+      addressLine2: Option[String],
+      city: String,
+      region: Option[String],
+      countryCode: String,
+      postCode: String
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/forms/ImporterManualAddressFormProvider.scala:40]: Same

File: app/models/Address.scala [hunk 7]
```diff
@@ -16,35 +16,20 @@
 
 package models
 
-import models.responses.LookedUpAddressWrapper
 import play.api.libs.json.{Json, OFormat}
 
+// TODO - add audit ref to this class (for ALF)
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/models/Address.scala:21]: Good question.  I'll add the field to the Address model.  Populate it from the ALF response and see where that takes us re: auditing.

[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/models/Address.scala]: I've added the auditRef field to the postcode and ensured that it is preserved when submitting our manual address page if the address hasn't changed.

But for the reference to be audited I think that we need a new ticket to add the field to the Address payload in the BE and from there copy it to the audit model.

File: app/services/AddressLookupService.scala [hunk 8]
```diff
@@ -40,16 +40,20 @@ class AddressLookupService @Inject() (
...
     lookupPageHeadingKey: String,
-    hintKey: String
+    hintKey: String,
+    editPageHeadingKey: String,
+    confirmationHeadingKey: String
   )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AddressLookupOnRamp] =
     addressLookupConnector.initialiseJourney(
       AddressLookupRequest(
-        appConfig.selfUrl(callBackUrl),
+        appConfig.selfUrl(callBackUrl.split('?').head),
```
Comment (njmoorewitchcat) [inline app/services/AddressLookupService.scala:49]: AMBER: I didn't think we where manipulating the URL in the service anymore - we would take what the Controller passed us (so we could add additional parameters if necessary)

**Review by njmoorewitchcat — APPROVED** [review]

File: app/services/AddressLookupService.scala (hunk 8 above)
Comment (SteveSugden) [reply app/services/AddressLookupService.scala]: I must have "prototyped" this fix on NDRC and then forgot to remove it.   Removed now.

**Pull request merged**